### PR TITLE
Generate and Run batch jobs for the Bird Dataset

### PR DESCRIPTION
**Pull request by fi5421**
## Description

This PR corresponds to the following [Task](https://www.notion.so/conradlabshq/Create-Batch-Jobs-for-the-Bird-Dataset-06c8c5f921db4c72b22c706fa6a0d3ef?pvs=4)

Made changes in the code to add a Sample_questions_directory variable in the config file that can be changed in the set schema function. vectorize from vectorize.py now uses this variable to fetch sample questions. Also changed the changeSchema endpoint to add an optional sample_path field that is used to set the sample_questions_directory. 

Made 2 scripts:

1. generate_batch_file.py script creates prompts, makes input batch files and creates batch jobs
2. download_batch_output.py script downloads the output from completed batch jobs by fetching batch ids from a file saved by generate_batch_file.py.

Note: Batch outputs are currently saved inside the DB directory and each batch job corresponds to a separate DB.

## For Reviewers 

Approve Plez:

image

### CODE REVIEW

File: server/scripts/batch_jobs_created.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+{"batch_67222597566c8190bca381d82769a175": "../data/bird/debit_card_specializing/test/", "batch_671f2c1a54f4819097fc7c9aa6ea4df2": "../data/bird/financial/test/"}
```
Comment (AwaisKamran) [inline server/scripts/batch_jobs_created.txt:1]: Why dont we keep it as a environment variable?

File: server/scripts/download_batch_output.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import os
+import json
+
+
+load_dotenv()
+OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
+
+openAI_client=OpenAI(api_key=OPENAI_API_KEY)
+
+time.sleep(5)
```
Comment (AwaisKamran) [inline server/scripts/download_batch_output.py:14]: Woah! Why?

File: server/scripts/download_batch_output.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    batch_jobs = json.loads(file.read())
+    file.close()
+
+count=0
+while len(downloaded)<len(batch_jobs):
+    print(f'Try: {count}')
+    for i in batch_jobs:
+        if  i not in downloaded:
+            job=openAI_client.batches.retrieve(i)
+            if job.status=='completed':
```
Comment (AwaisKamran) [inline server/scripts/download_batch_output.py:27]: Any such status should reside as an enum or a constant

File: server/scripts/generate_batch_file.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+MAX_TOKENS=1000
+TEMPERATURE=0.7
+
+
+script_path = "../data/bird/"
+relative_path='./data/bird/'
+db_change_url="http://localhost:8000/database/change/"
+prompt_generate_url='http://localhost:8000/prompts/generate/'
+prompt_type='dail_sql'
+num_shots=3
```
Comment (AwaisKamran) [inline server/scripts/generate_batch_file.py:19]: Should be in constants

File: server/scripts/download_batch_output.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+downloaded=[]
+
+with open('batch_jobs_created.txt', 'r') as file:
+    batch_jobs = json.loads(file.read())
+    file.close()
+
+count=0
+while len(downloaded)<len(batch_jobs):
+    print(f'Try: {count}')
+    for i in batch_jobs:
```
Comment (AwaisKamran) [inline server/scripts/download_batch_output.py:24]: Also, can you add a few comments on what this logic is doing?

File: server/scripts/generate_batch_file.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+directories = [d for d in os.listdir(script_path) if os.path.isdir(os.path.join(script_path, d))]
+directories=directories[:2]
```
Comment (AwaisKamran) [inline server/scripts/generate_batch_file.py:22]: Add a few comments here as well

File: server/scripts/generate_batch_file.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+                    ],
+                    "max_tokens": MAX_TOKENS,
+                    "temperature": TEMPERATURE
+                }
+            })
+    with open(f"{script_path}{i}/test/{i}_prompts_batch.jsonl",'w') as file:
+        for item in prompts:
+            json.dump(item, file)
+            file.write('\n')
+        file.close()
```
Comment (AwaisKamran) [inline server/scripts/generate_batch_file.py:51]: Comment this chunk out too, any function that exceeds 40 lines should be minimized by creating functions, like i see we can create a method for sending out API request

File: server/scripts/download_batch_output.py
Comment (Mehak-Conrad) [inline server/scripts/download_batch_output.py:1]: Repetetive code, should update and use : def download_file code in services/openai_client.py

File: server/scripts/download_batch_output.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        if  i not in downloaded:
+            job=openAI_client.batches.retrieve(i)
```
Comment (Mehak-Conrad) [inline server/scripts/download_batch_output.py:26]: Repetetive code, can update and use : def download_file code in services/openai_client.py

File: server/scripts/generate_batch_file.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+prompt_generate_url='http://localhost:8000/prompts/generate/'
+prompt_type='dail_sql'
+num_shots=3
+
+directories = [d for d in os.listdir(script_path) if os.path.isdir(os.path.join(script_path, d))]
+directories=directories[:2]
+
+for i in tqdm(directories,desc=f'Processing Directories'):
+
+    response=requests.post(db_change_url,json={'database_type':"hotel","sample_path":f"{relative_path}{i}/sample_questions/{i}_schema.json"})
```
Comment (Mehak-Conrad) [inline server/scripts/generate_batch_file.py:26]: Same review, repetitive code. can make a wrapper for create_batch_input_file() and update the file path only.

File: server/scripts/generate_batch_file.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+load_dotenv()
+OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
+
+openAI_client=OpenAI(api_key=OPENAI_API_KEY)
+uploaded_files=[]
+batch_jobs=[]
+batch_jobs_dict={}
+
+
+for i in tqdm(directories,desc='Uploading and Creating Batch Jobs'):
```
Comment (Mehak-Conrad) [inline server/scripts/generate_batch_file.py:65]: Look this up in openai client. create_batch_job, upload_batch_input_file, get_all_batches, get_all_uploaded_files, download_file functions are created.

**Review by Mehak-Conrad — APPROVED** [review]

**Pull request merged**